specification.md

```diff
@@ -0,0 +1,257 @@
+#   make build WHAT=cmd/server GOFLAGS=-v
+#   make all GOGCFLAGS="-N -l"
+#     Note: Use the -N -l options to disable compiler optimizations an inlining.
+#           Using these build options allows you to subsequently use source
+#           debugging tools like delve.
+```
+
+### Color log
+
+The Makefile SHOULD print colorized log when `$COLOR_LOG == true`
+
+### Version
+
+The valid version MUST be able to describe the current status of git tree.
+
+A complete version information contains:
+
+-   a git tag versioned by [Semantic Versioning 2.0.0](http://semver.org/spec/v2.0.0.html) (semver). e.g. v1.0.0 v0.1.0-alpha.1
+-   commits number after latest git tag and latest commit hash
+-   git tree status: 
+    -   `clean` indicates no changes since the git commit id.
+    -   `dirty` indicates source code changes after the git commit id
+    -   `archive` indicates the tree was produced by `git archive`
+
+#### Use Case
+
+| version                 | docker tag              | description                              |
+| ----------------------- | ----------------------- | ---------------------------------------- |
+| v0.0.1                  | v0.0.1                  | a clean git tree with the latest tag `v0.0.1` |
+| v0.0.1-dirty            | v0.0.1-dirty            | a dirty git tree with changes not staged |
+| v0.0.1-2+1b4531e1acf800 | v0.0.1-2-1b4531e1acf800 | a clean git tree with a tag ` v0.0.1`.<br />There are two new commits after the tag and the latest commit is `1b4531e1acf800` |
+
+If your git repo contains changes not staged, a `-dirty` suffix MUST always be appended to final version.
+
+#### Options
+
+| Name    | Usage                     | Type   | Default | Severity |
+| ------- | ------------------------- | ------ | ------- | -------- |
+| VERSION | used to replace `git tag` | string |         | REQUIRED |
+
+#### Example
+
+```bash
+VERSION=v0.0.1 make all
+```
+
+### Compile
+
+The Makefile MUST support the following targets for compiling.
+
+```bash
+make all # equal to make build
+make build
+make build-local
+make build-in-container
+```
+
+`make all` MUST be equal to `make build`
+
+`make build` means to compile you project. The Makefile MUST support the ability to compile the project on localhost and in docker container.
+
+ For convenience, the Makefile SHOULD support the ability to build targets directories directly.
+
+```bash
+make cmd/controller # equal to make build WHAT=cmd/controller
+```
+
+#### Options
+
+| Name        | Usage                                    | Type   | Default | Severity |
+| ----------- | ---------------------------------------- | ------ | ------- | -------- |
+| LOCAL_BUILD | If set true, project will be built on local machine, otherwise, built in docker. | string | false   | REQUIRED |
+| WHAT        | Directory names to build.  If any of these directories has a main package, the build will produce executable files under bin/. e.g `cmd/controller` | string |         | REQUIRED |
```

Comment:
is `WHAT` required?  looks like you don't have to use `WHAT`

Reply:
`make cmd/controller` is not Mandatory requirement. 
There MUST be a `WHAT` there to pass the directory names variables to make rule.

the most basic build rule is
```
make build WHAT=cmd/controller LOCAL_BUILD=true
```

the followings are extensions for the basic way
```
make build-local WHAT=cmd/controller
make build-in-linux WHAT=cmd/controller
make cmd/controller
make cmd/controller LOCAL_BUILD=false
```

Reply:
OK, LGTM.  I mis-read the meaning of `WHAT`.  I thought it means use must use `WHAT` option, but what required here means implementation must support the option.